### ISSUE TITLE
Inferred type with numpy is incorrect with pyright, correct with numpy

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#4081]
**Describe the bug**
pyright is not taking a type declaration of an `npt.NDArray[subtype]` correctly

**To Reproduce**
```python
from typing_extensions import reveal_type
import numpy as np
import numpy.typing as npt


ndarray_td64: npt.NDArray[np.timedelta64] = np.array([1, 2, 3], dtype="timedelta64[D]")

reveal_type(ndarray_td64)

foo: npt.NDArray[np.int_] = np.array([1,2,3])

reveal_type(foo)
```

pyright reports:
```text
  nparr.py:8:13 - information: Type of "ndarray_td64" is "ndarray[Any, dtype[Any]]"
  nparr.py:12:13 - information: Type of "foo" is "ndarray[Any, dtype[Any]]"
```

**Expected behavior**
mypy reports
```text
nparr.py:8: note: Revealed type is "numpy.ndarray[Any, numpy.dtype[numpy.timedelta64]]"
nparr.py:12: note: Revealed type is "numpy.ndarray[Any, numpy.dtype[numpy.signedinteger[Any]]]"
```

**Screenshots or Code**
See above.

Using numpy 1.23.3

**VS Code extension or command-line**
command line pyright 1.1.276

**Comment by erictraut** [comment]
This is related to an unspecified area of the Python type system — how to perform assignment-based type narrowing in the face of `Any` type arguments. Because this is undefined, pyright's current behavior is arguably correct, but I agree that mypy's behavior here is both more consistent and what I would expect as a user.

Here's a simplified code sample that exhibits the problem.

```python
from typing import Any, Generic, TypeVar

T = TypeVar("T")
class A(Generic[T]): ...

def func1(val: A[Any]):
    x: A[int] = val
    reveal_type(x) # A[int]

def func2(val: A[list[Any]]):
    x: A[list[int]] = val
    reveal_type(x) # A[list[Any]] rather than A[list[int]]
```

Pyright narrows the type as expected in `func1`, but it inconsistently deviates from this in `func2`. That's because the `Any` was located in a type argument to a type argument. 

I've updated the logic in pyright to handle this in a recursive manner for consistency.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.277, which I just published. It will also be included in a future release of pylance.
